### PR TITLE
[NFC] Refactor ChildLocalizer to handle unreachable code better

### PR DESCRIPTION
**Pull request by kripken**
This is NFC in the current users, but is necessary functionality for a later
PR.

ChildLocalizer moves children into locals as needed. It used to stop when it
saw the first unreachable. After this change we move such unreachable
children out of the parent as well, making this more uniform: all interacting
effects are moved out, and all that is left nested in the parent can be
moved around and removed as desired.

Also add a `getReplacement` helper that makes using this easier.

This cannot be tested comprehensively with the current user as that user
will not call this code path on an unreachable parent at all, so this just
adds what can be tested. The later PR will have tests for all corner cases
:crossed_fingers:

### CODE REVIEW

File: src/ir/localize.h [hunk 1]
```diff
@@ -106,6 +137,36 @@ struct ChildLocalizer {
...
+    if (sets.empty()) {
+      // Nothing to add.
+      return parent;
+    }
+
+    auto* block = Builder(wasm).makeBlock();
+    block->list.set(sets);
+    if (hasUnreachableChild) {
+      // If there is an unreachable child then we do not need the parent at all,
+      // and we know the type is unreachable.
```
Comment (tlively) [inline src/ir/localize.h:154]: What if the parent is needed for validation like the post-unreachable children? For example it could be a `br_if` where the condition is unreachable but where it is the only reason its target block has type none.

File: test/lit/passes/gto-removals.wast [hunk 2]
```diff
@@ -925,3 +925,49 @@
...
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (f64.const 30)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (struct.new_default $struct)
+  ;; CHECK-NEXT: )
+  (func $func (result (ref $struct))
+    ;; The fields can be removed here, but the effects must be preserved before
+    ;; the struct.new. The consts in the middle can vanish entirely.
```
Comment (tlively) [inline test/lit/passes/gto-removals.wast:957]: Why do we try to handle unreachability at all in GTO when a --dce before hand would have the same effect?

File: src/ir/localize.h (hunk 1 above)
Comment (kripken) [reply src/ir/localize.h:154]: Good point. That is not an issue atm as it is called on things that do not have that behavior, but the user needs to be aware of that. I'll add a note.

Comment (kripken) [reply src/ir/localize.h:161]: (added)

Comment (tlively) [reply src/ir/localize.h:161]: Would it be safer to just include the parent and let later DCE clean it up?

Comment (kripken) [reply src/ir/localize.h:161]: The caller needs to remove this code in some cases, and doing it there would be parsing the output from here to see if it is a block, and then removing checking if there is an unreachable child, and then removing the last block element if so. Then seems more complex to me.

**Review by tlively — APPROVED** [review]
Makes sense that some users would need to write their own parent removal logic if we didn't do it here, and that seems worse.
